vulcanai/models/utils.py

```diff
@@ -4,13 +4,75 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from torch.autograd import Variable
+from torch.utils.data import DataLoader, TensorDataset
 
 import numpy as np
 import math
 from sklearn.metrics import confusion_matrix
 from sklearn.preprocessing import LabelBinarizer
 from collections import OrderedDict as odict
+from collections import defaultdict
 
+def get_probs(network, loader, index_to_iter, ls_feat_vals):
+    """
+    Returns probability for each object within loader based on output from training neural network
+    Parameters:
+        network : vulcan.model
+	    training vulcan network
+	loader : torch.dataloader
+	    dataloader containing validation set
+	index_to_iter : string
+	    feature to iterate through adjusting values
+	ls_feat_vals : list
+	    values to iterate through for feature in index_to_iter
+    Returns:
+	dct_scores : dictionary
+	    dictionary of scores
+    """
+    dct_scores = defaultdict()
+    for index in range(len(loader)):
+        dct_scores[index] = {}
+    for index in range(len(loader)):
+        nn_subjFeat = torch.Tensor(loader.dataset[index][0].numpy())
+        nn_subjLabel = torch.LongTensor(loader.dataset[index][1].float().numpy())
+        input_loader = DataLoader(TensorDataset(nn_subjFeat.unsqueeze(0), nn_subjLabel.unsqueeze(0)))
+        subjProb = network.forward_pass(data_loader=input_loader, transform_outputs=False)
+        subjProb = subjProb[0][1] * 100
+        subjProb = round(subjProb, 2)
+        dct_scores[index][nn_subjFeat[index_to_iter].item()] = subjProb
+        for newVal in ls_feat_vals:
+            if newVal != nn_subjFeat[index_to_iter].item():
+                nn_subjFeat[index_to_iter] = newVal
+                input_loader = DataLoader(TensorDataset(nn_subjFeat.unsqueeze(0), nn_subjLabel.unsqueeze(0)))
+                subjProb = network.forward_pass(data_loader=input_loader, transform_outputs=False)
+                subjProb = subjProb[0][1] * 100
+                subjProb = round(subjProb, 2)
+                dct_scores[index][newVal] = subjProb
+    return dct_scores
+
+def filter_matched_subj(dct_scores, loader, index_to_iter):
+    """
+    Returns dictionary of filtered keys based on predicted value and value truly assigned in actual set.
+    Parameters:
+        dct_scores : dictionary
+            dictionary of probability scores produced in get_scores function
+        loader : torch.dataloader
+            dataloader containing validation set
+        index_to_iter : string
+            feature to iterate through adjusting values
+    Returns:
+        dct_filtered : dictionary
+            dictionary of filtered entities 
+    """
+    dct_filtered = {}
+    for subj in list(dct_scores):
+        highest_prob = max(dct_scores[subj].values())
+        highest_val = [val for val, prob in dct_scores[subj].items() if prob==highest_prob]
+        if len(highest_val) == len(list(dct_scores[subj])):
```

Comment:
this is if they're tied? just want to clarify my understanding

Reply:
what if there's like, half of them with the same top score? that situation isn't covered, unless it doesn't need to be?

Reply:
So the `elif` statement checks if there is multiple but not all with same top score. `highest_val` is a list so `elif` checks if the value in the original dataset is in the `highest_val` list.

Reply:
I'm still confused, because isn't the  `if` checking for exact equivalence in terms of list length? I understand that you want to get the probability value if it's the highest for your target treatment.... but I don't understand why the first `if` is there? sorry...

Reply:
Actually, looking at it again we can remove the `if` statement and just have the `elif` be the only check. Because all we care about is if the value assigned to the subject is within the list of `highest_prob` because `highest_prob` is the list of values that have the highest probability.

Reply:
ok cool, that resolves my confusion.